esp-hal/src/uart.rs

```diff
@@ -510,27 +461,76 @@ where
     }
 }
 
+impl<'d, T> UartRx<'d, T, Blocking>
+where
+    T: Instance + 'd,
+{
+    /// Create a new UART RX instance in [`Blocking`] mode.
+    pub fn new<RX: InputPin>(
+        _uart: impl Peripheral<P = T> + 'd,
+        rx: impl Peripheral<P = RX> + 'd,
+    ) -> Self {
+        crate::into_ref!(rx);
+        rx.set_to_input(crate::private::Internal);
+        rx.connect_input_to_peripheral(T::rx_signal(), crate::private::Internal);
+        Self::new_inner()
+    }
+}
+
 impl<'d, T> Uart<'d, T, Blocking>
 where
     T: Instance + 'd,
 {
     /// Create a new UART instance with configuration options in [`Blocking`]
     /// mode.
-    pub fn new_with_config<P>(
+    pub fn new_with_config<TX: OutputPin, RX: InputPin>(
         uart: impl Peripheral<P = T> + 'd,
         config: Config,
-        pins: Option<P>,
         clocks: &Clocks,
         interrupt: Option<InterruptHandler>,
-    ) -> Self
-    where
-        P: UartPins,
-    {
-        Self::new_with_config_inner(uart, config, pins, clocks, interrupt)
+        tx: impl Peripheral<P = TX> + 'd,
+        rx: impl Peripheral<P = RX> + 'd,
+    ) -> Self {
+        crate::into_ref!(tx);
+        crate::into_ref!(rx);
+        tx.set_to_push_pull_output(crate::private::Internal);
+        tx.connect_peripheral_to_output(T::tx_signal(), crate::private::Internal);
+
+        rx.set_to_input(crate::private::Internal);
+        rx.connect_input_to_peripheral(T::rx_signal(), crate::private::Internal);
+        Self::new_with_config_inner(uart, config, clocks, interrupt)
+    }
+
+    /// Create a new UART instance with defaults in [`Blocking`] mode.
+    pub fn new<TX: OutputPin, RX: InputPin>(
+        uart: impl Peripheral<P = T> + 'd,
+        clocks: &Clocks,
+        tx: impl Peripheral<P = TX> + 'd,
+        rx: impl Peripheral<P = RX> + 'd,
+    ) -> Self {
+        crate::into_ref!(tx);
+        crate::into_ref!(rx);
+        tx.set_to_push_pull_output(crate::private::Internal);
+        tx.connect_peripheral_to_output(T::tx_signal(), crate::private::Internal);
+
+        rx.set_to_input(crate::private::Internal);
+        rx.connect_input_to_peripheral(T::rx_signal(), crate::private::Internal);
+        Self::new_inner(uart, clocks)
     }
 
     /// Create a new UART instance with defaults in [`Blocking`] mode.
-    pub fn new(uart: impl Peripheral<P = T> + 'd, clocks: &Clocks) -> Self {
+    /// Verify that the default pins (DefaultTxPin and DefaultRxPin) are used.
+    pub fn new_with_default_pins(
```

Comment:
I'm not sure I understand this constructor, how is this different from just calling `new`?

Reply:
Oh wait I'm just confused I think, there are way too many constructors and not enough documentation. Disregard I guess.